envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/ServiceTagsAndCanaryTest.kt

```diff
@@ -0,0 +1,155 @@
+package pl.allegro.tech.servicemesh.envoycontrol
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.AfterAll
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlRunnerTestApp
+import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlTestConfiguration
+import pl.allegro.tech.servicemesh.envoycontrol.config.containers.EchoContainer
+
+open class ServiceTagsAndCanaryTest : EnvoyControlTestConfiguration() {
+
+    companion object {
+        private val properties = mapOf(
+            "envoy-control.envoy.snapshot.routing.service-tags.enabled" to true,
+            "envoy-control.envoy.snapshot.routing.service-tags.metadata-key" to "tag",
+            "envoy-control.envoy.snapshot.load-balancing.canary.enabled" to true,
+            "envoy-control.envoy.snapshot.load-balancing.canary.metadata-key" to "canary",
+            "envoy-control.envoy.snapshot.load-balancing.canary.metadata-value" to "1"
+        )
+
+        @JvmStatic
+        @BeforeAll
+        fun setupTest() {
+            setup(appFactoryForEc1 = { consulPort ->
+                EnvoyControlRunnerTestApp(properties = properties, consulPort = consulPort)
+            })
+        }
+
+        private val loremRegularContainer = echoContainer
+        private val loremCanaryContainer = echoContainer2
+        private val ipsumRegularContainer = EchoContainer().also { it.start() }
+
+        @JvmStatic
+        @AfterAll
+        fun cleanup() {
+            ipsumRegularContainer.stop()
+        }
+    }
+
+    protected fun registerServices() {
+        registerService(name = "echo", container = loremRegularContainer, tags = listOf("lorem"))
+        registerService(name = "echo", container = loremCanaryContainer, tags = listOf("lorem", "canary"))
+        registerService(name = "echo", container = ipsumRegularContainer, tags = listOf("ipsum"))
+    }
+
+    @Test
+    fun `should route requests to canary instance with tag lorem`() {
+        // given
+        registerServices()
+        untilAsserted {
```

Comment:
Definitely dry it out, it appears in every test. I think there is a helper function somewhere. Please check.

Reply:
Yeah, I added the helper function for that in this PR but then i didn't use it in all places :) Fixed